marketplace/google-analytics-app/src/AppConfig.tsx

```diff
@@ -0,0 +1,302 @@
+import * as React from 'react';
+import {
+  Typography,
+  Heading,
+  Paragraph,
+  TextField,
+  Option,
+  TextLink,
+  Button,
+  Select,
+  FormLabel,
+  TextInput
+} from '@contentful/forma-36-react-components';
+import styles from './styles';
+import PropTypes from 'prop-types';
+import { AppConfigParams, AppConfigState } from './typings';
+import { getSidebarLocations } from './utils';
+
+export default class AppConfig extends React.Component<AppConfigParams, AppConfigState> {
+  static propTypes = {
+    sdk: PropTypes.object.isRequired
+  };
+
+  state: AppConfigState = {
+    allContentTypes: {},
+    contentTypes: {}
+  };
+
+  async componentDidMount() {
+    const { sdk } = this.props;
+    const { app } = sdk.platformAlpha;
+
+    const [{ items: spaceContentTypes }, savedParams] = await Promise.all([
+      sdk.space.getContentTypes().then(),
+      app.getParameters()
+    ]);
+
+    const sidebarLocations = await getSidebarLocations(sdk);
+    // remove content types for which the app has been removed from the sidebar
+    const contentTypes = sidebarLocations.reduce((acc, key) => {
+      const saved = savedParams.contentTypes[key];
+
+      if (key && saved) {
+        acc[key] = saved;
+      }
+
+      return acc;
+    }, {});
+
+    // add an incomplete contentType entry if there are none saved
+    if (!Object.keys(contentTypes).length) {
+      contentTypes[''] = { slugField: '', urlPrefix: '' };
+    }
+
+    // eslint-disable-next-line react/no-did-mount-set-state
+    this.setState(
+      {
+        allContentTypes: spaceContentTypes.reduce((acc, contentType) => {
+          contentType.fields = contentType.fields
+            // use only short text fields of content type
+            .filter(f => f.type === 'Symbol')
+            // sort by field name
+            .sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
+          acc[contentType.sys.id] = contentType;
+
+          return acc;
+        }, {}),
+        contentTypes,
+        clientId: savedParams.clientId || '',
+        viewId: savedParams.viewId || ''
+      },
+      () => sdk.app.setReady()
+    );
+
+    app.onConfigure(() => this.configureApp());
+  }
+
+  async configureApp() {
+    const { contentTypes, clientId, viewId } = this.state;
+
+    const EditorInterface = Object.keys(contentTypes).reduce((acc, id) => {
+      acc[id] = { sidebar: { position: 0 } };
+      return acc;
+    }, {});
+
+    return {
+      parameters: {
+        contentTypes,
+        clientId,
+        viewId
+      },
+      targetState: {
+        EditorInterface
+      }
+    };
+  }
+
+  handleChange(name, value) {
+    this.setState({ [name]: value });
+  }
+
+  handleContentTypeChange(prevKey, key) {
+    this.setState(prevState => {
+      const contentTypes = {};
+
+      // remove contentType[prevKey] field and replace with the new contentType
+      // key while preserving key order
```

Comment:
Not sure what you mean by this. We shouldn't rely on keys in an object to preserve any order. If we need order, we should use an array instead.

Reply:
This is only to avoid the order of the rows of content types changing because of a change in the output `Object.keys()` when the user selects a different content type for a row. I could this properly by maintaining an array of selected keys, but this is much easier and only breaks with all-number keys.